### DISCUSSION TITLE
Current password requirements degrade security

### DISCUSSION THREAD

**Discussion by rakleed** [rustdesk/rustdesk#2888] [General]
**Describe the bug you encountered:**

The current password requirements make security worse as I can't make a 3-4 word password separated by spaces (eg `holt pimple unstop warlord`). And when using numbers and capital letters, it will be more difficult for me to remember the password, although from a security point of view, the length of the password is more important in this case.

**What did you expect to happen instead?**

With a password length of 16 characters or more, requirements with at least one digit and a capital letter should be ignored. More: https://security.stackexchange.com/q/6095 and [this one](https://1password.com/password-generator/#:~:text=I%20need%20to%20generate%20a%20memorable%2C%20but%20strong%20password.%20Any%20tips%3F).

**How did you install `RustDesk`?**

GitHub release nightly

---

**RustDesk version and environment**

RustDesk 1.2.0 build date 2023-01-20 02:36
Windows 11 v. 22H2/macOS 13.1

**Comment by rakleed** [comment]
@21pages

**Comment by rakleed** [comment]
https://github.com/rustdesk/rustdesk/commit/9aecd287028e92f73d861002b5abcb88ca6be85e#diff-522753da202a73e97779ec07417bf4e3e7df19d0771f6f914874fa7c45fe724d

**Comment by working-name** [comment]
So what does it mean? Will you or will you not remove the silly password requirements? If not, can you make them optional? `wontfix` label is confusing.

**Comment by 21pages** [comment]
Glad to see a better way, and it should be simple.

**Reply by rakleed** [reply]
What do you think about such an option?

> With a password length of 16 characters or more, requirements with at least one digit and a capital letter should be ignored.

**Reply by working-name** [reply]
A better way would be to allow the user to make choices. Yes, warn them in big red letters that it's insecure, but ultimately it's their choice.

**Reply by pedjas** [reply]
> A better way would be to allow the user to make choices. Yes, warn them in big red letters that it's insecure, but ultimately it's their choice.

I agree. Enforcing any rules of this kind to users is problematic. Let users deal with it.

**Reply by spiderfudge** [reply]
Agreed!

**Comment by rawiriblundell** [comment]
Bump.  First time trying rustdesk today and I'm a little disappointed with the password policy restriction.

NIST is the gold standard that everyone else references:

https://pages.nist.gov/800-63-3/sp800-63b.html#memsecretver

Noting:

>Verifiers SHOULD NOT impose other composition rules (e.g., requiring mixtures of different character types or prohibiting consecutively repeated characters) for memorized secrets.

>“Verifiers SHOULD NOT require memorized secrets to be changed arbitrarily (e.g., periodically). However, verifiers SHALL force a change if there is evidence of compromise of the authenticator.”

That spec is quite dry reading, so you can also refer to e.g. OWASP's cheat sheets:

https://cheatsheetseries.owasp.org/cheatsheets/Authentication_Cheat_Sheet.html

**Reply by working-name** [reply]
Come on @rustdesk, just let this die, it's a bad idea. 

A password that's exponentially harder to crack is `I wish rustdesk would get off my back.` as opposed to `myPass00`. Somehow the first one isn't going to be secure according to the current requirements 🤷‍♂️

**Comment by rustdesk** [comment]
> NIST is the gold standard that everyone else references:
> 
> https://pages.nist.gov/800-63-3/sp800-63b.html#memsecretver
> 
> Noting:
> 
> > Verifiers SHOULD NOT impose other composition rules (e.g., requiring mixtures of different character types or prohibiting consecutively repeated characters) for memorized secrets.
> 
> > “Verifiers SHOULD NOT require memorized secrets to be changed arbitrarily (e.g., periodically). However, verifiers SHALL force a change if there is evidence of compromise of the authenticator.”
> 
> That spec is quite dry reading, so you can also refer to e.g. OWASP's cheat sheets:
> 
> https://cheatsheetseries.owasp.org/cheatsheets/Authentication_Cheat_Sheet.html

@sahilyeole check above two links.

Strong password recommendation from Avast:
https://www.avast.com/en-sg/random-password-generator

![image](https://github.com/rustdesk/rustdesk/assets/71636191/8fc689f9-54a8-4d23-9a4f-7d0d229c6f40)

**Reply by pedjas** [reply]
That is recommendation, That means it is not enforced rule. And it is on user to decide what password to use.

**Reply by parcelcat** [reply]
It would be a mistake to rely on Avast's recommendation. [Avast was fined $16.5 million](https://www.pcmag.com/news/ftc-bans-antivirus-provider-avast-from-selling-users-browsing-data) a couple of years ago for selling user data collected from its antivirus software to third party companies, and banned from doing it again. Not exactly a trustworthy actor.

What everyone else here is saying is much more reasonable than what Avast has to say about security.

**Comment by Boxylmer** [comment]
Chiming in with similar thoughts after upgrading and finding this out, using RustDesk on an internal network and also using `correct` `horse` `battery` `staple` like passwords. Can this not at least be a server option? Typically users hosting RustDesk relays are at least somewhat aware of secure practices.

**Comment by godzfire** [comment]
Updates to this? I have a specific very secure password, but it doesn't have a capital, so as it stands I can't use it. As other state forcing password restrictions isn't up to Rustdesk, it's up to the user. It should be changed to warn the user if the regular requirements are not met, but not stop them.

**Reply by Boxylmer** [reply]
How difficult you think it'd be to fork the package each major release and just make the change? The devs seem unwilling to at least add a server flag.

**Comment by zjcurtis** [comment]
Yeah I would be very interested in seeing if this is something the team is open to fixing, not fixing something that has been raised as counter to good security standards is concerning.

**Comment by LoganDark** [comment]
Since #10179 was apparently closed as duplicate of this, what is RustDesk's stance on this? I am also running into the "missing uppercase letter in an otherwise very secure password" problem and have received no actual comment on my issue

**Comment by zjcurtis** [comment]
The position expressed by RustDesk on Sep 8 2023 was that Avast suggested using complexity as a security factor, so (seemingly) Rustdesk should keep doing it. I'd like to show that this is, _**at least**_ up for debate.

- [NIST](https://pages.nist.gov/800-63-4/sp800-63b/passwords/#complexity) has been brought up.

- [CIS Password Policy Guide](https://www.cisecurity.org/insights/white-papers/cis-password-policy-guide) excerpting from the guide in section 5.1.3 "Allow the use of any type character, and limit enforcing specific characters types"

- [Australian Signals Directorate](https://www.cyber.gov.au/protect-yourself/securing-your-accounts/passphrases) recommending passphrases over passwords.

- [UK National Cyber Security Centre](https://www.ncsc.gov.uk/collection/top-tips-for-staying-secure-online/three-random-words) recommending (relatively short) passphrases and explicitly suggesting that swapping in numeric characters into a passphrase is not a worthwhile trade. 

Here are the options as I see them:

- Change the complexity requirement
- Keep the requirement
- Make it configurable with or without some recommended setting.
- Allow authentication with just a passkey.

ISO 27002 still seems to suggest having some sort of complexity standard. Having this be configurable would allow a variety of users, under different compliance/regulatory regimes, to use RustDesk.

**Reply by Boxylmer** [reply]
Configuration is what I believe most here are advocating for.

**Reply by LoganDark** [reply]
> The position expressed by RustDesk on Sep 8 2023 was that Avast suggested using complexity as a security factor

All that was expressed explicitly is that Avast says otherwise, although yes, that does seem to be the implication. I don't trust Avast though, they're biased (they have an incentive to give you as many security guidelines as possible, because they're an antivirus product). From the perspective of the designer of an application, there's no reason to require every single thing that they say is good practice - I'd honestly prefer if RustDesk just used [zxcvbn](https://github.com/dropbox/zxcvbn) or something instead of trying to decide on which hard requirements make a password secure. Under this model, you can go above and beyond every single requirement but a single one and then still be blocked from setting the password, not because it's in any way insecure, but because "yOuR pAsSwOrD mUsT cOnTaIn aN uPpErCaSe lEtTeR".

**Reply by rawiriblundell** [reply]
Since my last comment, NIST have released updated guidelines.  A lot has changed because it's a vastly comprehensive spec, but the TL;DR is that a bunch of items have been moved from "SHOULD" to "SHALL", which in plain English means "from _recommended_ to _requirement_".

I want to acknowledge that @zjcurtis posted a link to the updated version.  In addition, I'd like to throw in this one:

https://pages.nist.gov/800-63-4/sp800-63b.html#passwordver

Which more directly lays out NIST's position vs the old link I provided:

https://pages.nist.gov/800-63-3/sp800-63b.html#memsecretver

Noting, also, the context stated in the [change log](https://pages.nist.gov/800-63-4/sp800-63b.html#s-d) that "memorized secrets" are now referred to as "passwords".

To copy and paste the juicy bits:

> 1. Verifiers and CSPs SHALL require passwords to be a minimum of eight characters in length and SHOULD require passwords to be a minimum of 15 characters in length.
> 2. Verifiers and CSPs SHOULD permit a maximum password length of at least 64 characters.
> 3. Verifiers and CSPs SHOULD accept all printing ASCII [[RFC20]](https://pages.nist.gov/800-63-4/sp800-63b.html#ref-RFC20) characters and the space character in passwords.
> 4. Verifiers and CSPs SHOULD accept Unicode [[ISO/ISC 10646]](https://pages.nist.gov/800-63-4/sp800-63b.html#ref-ISOIEC10646) characters in passwords. Each Unicode code point SHALL be counted as a single character when evaluating password length.
> 5. Verifiers and CSPs SHALL NOT impose other composition rules (e.g., requiring mixtures of different character types) for passwords.
> 6. Verifiers and CSPs SHALL NOT require users to change passwords periodically. However, verifiers SHALL force a change if there is evidence of compromise of the authenticator.
> 7. Verifiers and CSPs SHALL NOT permit the subscriber to store a hint that is accessible to an unauthenticated claimant.
> 8. Verifiers and CSPs SHALL NOT prompt subscribers to use knowledge-based authentication (KBA) (e.g., “What was the name of your first pet?”) or security questions when choosing passwords.
> 9. Verifiers SHALL verify the entire submitted password (i.e., not truncate it).

The most important and relevant part is as follows:

Old version:

>Verifiers SHOULD NOT impose other composition rules (e.g., requiring mixtures of different character types or prohibiting consecutively repeated characters) for memorized secrets. Verifiers SHOULD NOT require memorized secrets to be changed arbitrarily (e.g., periodically). However, verifiers SHALL force a change if there is evidence of compromise of the authenticator.

New version:

> 5. Verifiers and CSPs SHALL NOT impose other composition rules (e.g., requiring mixtures of different character types) for passwords.
> 6. Verifiers and CSPs SHALL NOT require users to change passwords periodically. However, verifiers SHALL force a change if there is evidence of compromise of the authenticator.

So we can see that a couple of SHOULD NOT's (recommendations) have been changed to SHALL NOT's (requirements).

And this recommendation was present in the previous version and is slightly changed in the latest version:

>When processing a request to establish or change a password, verifiers SHALL compare the prospective secret against a blocklist that contains known commonly used, expected, or compromised passwords. The entire password SHALL be subject to comparison, not substrings or words that might be contained therein.

This supports @LoganDark's recommendation of  [zxcvbn](https://github.com/dropbox/zxcvbn), and I'll happily second that nomination.

Obviously there's no law that says that one must be in full 100% compliance with NIST.  I want to re-state that they are, nonetheless, the gold standard that everybody else ultimately references.  So when NIST makes a change, everybody else is likely to review and update their standards in the following months, if not years.  Their standards are ultimately a good indicator of the direction things are heading.

**Reply by parcelcat** [reply]
> I don't trust Avast though, they're biased

Avast isn't just biased, they're [downright untrustworthy](https://www.pcmag.com/news/ftc-bans-antivirus-provider-avast-from-selling-users-browsing-data). I'm disappointed that RustDesk is relying on them for security guidance.

**Comment by jgoldgh** [comment]
I agree that having these restrictions on passwords is a combination of stupidity and arrogance on the part of the developers.  I'm perfectly capable of coming up with my own passwords. I know my platform. I know my requirements. I don't need some other person making those decisions from a position of ignorance.  Restrictions that limit the use of all password options for a given length make brute force attacks easier.

I'm just evaluating rustdesk. I've heard lots of good things. This is a big disappointment. Stop messing with people's passwords!

**Comment by MMagneson** [comment]
The capital letters and numbers/symbols requirement is a security anachronism from Windows NT in the early 2000's and should be abolished by now. 

When setting a password in the settings, Passw0rd! gets approved as a strong password(!) - which is one of the worst passwords you can have. This will according to security.org take about 3 weeks for a supercomputer to brute-force without a dictionary. 

Meanwhile "ejlikastæktekræde", a word in a norwegian dialect not possible to find in a dictionary, 17 characters long, and extremely easy for me to remember is refused as too weak. Still, according to security.org, this will take one hundred quadrillion years for a supercomputer to crack - and nobody will wait around for that to happen. (I don't, and you now shouldn't, use "ejlikastæktekræde" as a password).

A 16 character minimum, ditching the capital/numbers/specials would be a way better and safer requirement.

**Comment by fernandoslessa** [comment]
Why is it still a thing? Users should be responsible for defining a password and not an annoying password strength verification logic defined by the devs. I still use the version 1.1.9 because it is the last one that does not annoy me with password requirements or anything else, it is just simple and easy to use.

**Comment by TheQWERTYCodr** [comment]
The linked password generator from Avast (where you can find the outdated security guidance) doesn't even enable symbols by default.

**Comment by zeldatp151** [comment]
How has this issue not yet been fixed. I'm trying to make a secure password that can actually be remembered and not written down, but these password requirements for 1999 are preventing me from doing so.

**Comment by basenc** [comment]
cmon its one crate away from being good

**Comment by charles-monson-win** [comment]
For what it's worth to everyone using multi-word passphrases, I use Diceware passwords for anything I have to type in and I've long used a different separator than spaces to deal with password complexity requirements of various software/sites. 

So instead of:

`correct horse battery staple`

A different separator or capitalization can be used:

`correct-horse-battery-staple` or `CorrectHorseBatteryStaple`

To meet most complexity requirements various software seem to use I settled on using the capitalization separated words and appending a consistent suffix of numbers with special characters to my Diceware generated passwords.

Something like this:

`CorrectHorseBatteryStaple12#$` - Site1

`WrongDonkeyCellPaperl12#$` - Site2


Meets 99.9% of sites/software's complexity requirements, remains memorable since the separator and suffix I chose don't change, and stays as secure as a normal Diceware password without the suffix (slightly better I guess, since I don't tell anyone what the suffix is).

That all being said... complexity requirements should just be configurable like most software allows.
